client/src/main/java/com/scalar/dl/client/service/GenericContractClientService.java

```diff
@@ -454,9 +759,32 @@ private String buildAssetId(AssetType type, List<String> keys) {
       case COLLECTION:
         checkArgument(keys.size() == 1, ClientError.SERVICE_WRONG_KEYS_SPECIFIED.buildMessage());
         return COLLECTION_ID_PREFIX + keys.get(0);
+      case TABLE:
+        checkArgument(keys.size() == 1, ClientError.SERVICE_WRONG_KEYS_SPECIFIED.buildMessage());
+        return PREFIX_TABLE + keys.get(0);
+      case RECORD:
+        checkArgument(keys.size() == 3, ClientError.SERVICE_WRONG_KEYS_SPECIFIED.buildMessage());
+        return PREFIX_RECORD + String.join(ASSET_ID_SEPARATOR, keys);
+      case INDEX:
+        checkArgument(keys.size() == 3, ClientError.SERVICE_WRONG_KEYS_SPECIFIED.buildMessage());
+        return PREFIX_INDEX + String.join(ASSET_ID_SEPARATOR, keys);
       default:
         throw new IllegalArgumentException(
             ClientError.SERVICE_WRONG_ASSET_TYPE_SPECIFIED.buildMessage());
     }
   }
+
+  private String toStringFrom(JsonValue value) {
+    return toStringFrom(jacksonSerDe.deserialize(value.toString()));
+  }
+
+  private String toStringFrom(JsonNode value) {
```

Comment:
Since the numeric value conversion is based on the logic in the Jackson-based table generic contracts, we need to convert JsonObejct and String values to JsonNode ones to get correct asset IDs, although they sometimes lose precision; e.g., a large decimal value of JsonObject/String will be handled as double in Jackson. We should notice it in the documents somewhere.